src/pcre2_compile_class.c

```diff
@@ -911,14 +913,45 @@ restriction is in force). Sometimes we can just extend the original range. */
 if ((options & PCRE2_CASELESS) != 0)
   {
 #ifdef SUPPORT_UNICODE
-  if ((options & (PCRE2_UTF|PCRE2_UCP)) == 0)
+  /* UTF mode. This branch is taken if we don't support wide characters (e.g.
+  8-bit library, without UTF), but we do treat those characters as Unicode
+  (if UCP flag is set). In this case, we only need to expand the character class
+  set to include the case pairs which are in the 0-255 codepoint range. */
+  if ((options & (PCRE2_UTF|PCRE2_UCP)) != 0)
+    {
+      BOOL turkish_i = (xoptions & (PCRE2_EXTRA_TURKISH_CASING|PCRE2_EXTRA_CASELESS_RESTRICT)) ==
```

Comment:
Don't this code duplicate logic here? This should happen only when 8 bit mode, without uftf, with ucp.

Reply:
Oh ouch. This is seriously-complicated code, with tight coupling.

You're right, with my change, we are doing caselessness twice, in some conditions, like 16-bit-without-UTF.

I've added one small extra change in a commit I've just pushed. I've also experimented with some additional asserts, to check that casing isn't being done twice.

---

I think that if we put a 256-bit map at the start of the cranges, then we can stack-allocate it for the no-wide-chars condition. It would also make it a little more efficient for UTF patterns which are only using ASCII characters (no need to heap-allocate those either).

Reply:
I may misunderstand the code, but `(options & (PCRE2_UTF|PCRE2_UCP)) != 0` always redo the caseless computing, not just in non ucp mode.
What about checking the 8 bit with `#if` then use `if ((options & (PCRE2_UTF|PCRE2_UCP)) == PCRE2_UCP)` to limit

Reply:
The code in the PR passes all the tests with this additional assertion, below.

However, I'm not sure I want to actually add that assertion in the code, since then everything becomes even more tightly coupled.

In any case, your invariant is currently true that `(options & (PCRE2_UTF|PCRE2_UCP)) == PCRE2_UCP`.

```diff
diff --git a/src/pcre2_compile_class.c b/src/pcre2_compile_class.c
index 329c9e89..cb73eefb 100644
--- a/src/pcre2_compile_class.c
+++ b/src/pcre2_compile_class.c
@@ -909,30 +909,31 @@ uint32_t classbits_end = (end <= 0xff ? end : 0xff);
 cases. In Unicode, there are 8-bit characters that have alternate cases that
 are greater than 255 and vice-versa (though these may be ignored if caseless
 restriction is in force). Sometimes we can just extend the original range. */

 if ((options & PCRE2_CASELESS) != 0)
   {
 #ifdef SUPPORT_UNICODE
   /* UTF mode. This branch is taken if we don't support wide characters (e.g.
   8-bit library, without UTF), but we do treat those characters as Unicode
   (if UCP flag is set). In this case, we only need to expand the character class
   set to include the case pairs which are in the 0-255 codepoint range. */
   if ((options & (PCRE2_UTF|PCRE2_UCP)) != 0)
     {
+      PCRE2_ASSERT((options & (PCRE2_UTF|PCRE2_UCP)) == PCRE2_UCP);
       BOOL turkish_i = (xoptions & (PCRE2_EXTRA_TURKISH_CASING|PCRE2_EXTRA_CASELESS_RESTRICT)) ==
         PCRE2_EXTRA_TURKISH_CASING;
       if (start < 128)
         {
         uint32_t lo_end = (classbits_end < 127 ? classbits_end : 127);
         for (c = start; c <= lo_end; c++)
           {
           if (turkish_i && UCD_ANY_I(c)) continue;
           SETBIT(classbits, cb->fcc[c]);
           }
         }
       if (classbits_end >= 128)
         {
         uint32_t hi_start = (start > 128 ? start : 128);
         for (c = hi_start; c <= classbits_end; c++)
           {
           uint32_t co = UCD_OTHERCASE(c);
```